---
title: "Gaussian Processes Regression"
author: "Andrew Esch, Evan Lee, and Collin Stratton"
date: "2-27-2022"
output: html_notebook
---

# Introduction
TK

## Gene Expression and Differential Expressions
Gene Expression is the biological process found in cells in which genetic information (coded in the set of bases A, G, T, and C) is decoded and utilized to form a set of genetic products. The types of products that the process of gene expression can produce includes amino acids and proteins. This process is essential foundation for the formation of all biological life, and the differences in gene expression between cells is an important area of biology for further study.

Differential expression is the observation of a statistically significant difference or change in read counts or expression levels. To find a differential expression between genes, it is important to use statistical models to find if there is statistical significants for these metrics.


## Control and Treatment Categorizations
A control is a group categorization that does not have any affected conditions. In other words, this group acts as the baseline for all other measurements in a model. The standard for most models is to feature one control. Alternatively, a treatment is a type of group categorization that has been affected by some type of condition. They are also known as experimental group categorizations. A model may feature several treatments.


```{r}
#RUN ONCE THEN COMMENT OUT
# if (!requireNamespace("BiocManager"))
#    install.packages("BiocManager")
# BiocManager::install(c("limma", "edgeR", "Glimma", "org.Mm.eg.db", "gplots", "RColorBrewer", "NMF", "BiasedUrn"))

library(edgeR)
# edgeRUserGuide()
library(limma)
library(Glimma)
library(org.Mm.eg.db)
library(gplots)
library(RColorBrewer)
library(NMF)
library(MASS)
library(plyr)
library(reshape2)
library(ggplot2)
library(GauPro)
set.seed(1234)

```

```{r Import Data}

#Evan Import -> C:/GCU Academics/Junior Year/Second Semester/CST-425/CLC Shared GitHub Repository/Cloned Repository/GaussianProcessesRegression/

#Andrew Import ->

# Read the data into R
seqdata <- read.delim("C:/GCU Academics/Junior Year/Second Semester/CST-425/CLC Shared GitHub Repository/Cloned Repository/GaussianProcessesRegression/GSE60450_LactationGenewiseCounts.txt", stringsAsFactors = FALSE)
# Read the sample information into R
sampleinfo <- read.delim("C:/GCU Academics/Junior Year/Second Semester/CST-425/CLC Shared GitHub Repository/Cloned Repository/GaussianProcessesRegression/SampleInfo.txt", stringsAsFactors = TRUE)

head(seqdata)
dim(seqdata)

sampleinfo
```



```{r TEST}
library(MASS)
gaussprocess <- function(from = 0, to = 1, K = function(s, t) {min(s, t)},
                         start = 0, m = 1000) {
  # Simulates a Gaussian process with a given kernel
  #
  # args:
  #   from: numeric for the starting location of the sequence
  #   to: numeric for the ending location of the sequence
  #   K: a function that corresponds to the kernel (covariance function) of
  #      the process; must give numeric outputs, and if this won't produce a
  #      positive semi-definite matrix, it could fail; default is a Wiener
  #      process
  #   start: numeric for the starting position of the process
  #   m: positive integer for the number of points in the process to simulate
  #
  # return:
  #   A data.frame with variables "t" for the time index and "xt" for the value
  #   of the process

  t <- seq(from = from, to = to, length.out = m)
  Sigma <- sapply(t, function(s1) {
    sapply(t, function(s2) {
      K(s1, s2)
    })
  })

  path <- mvrnorm(mu = rep(0, times = m), Sigma = Sigma)
  path <- path - path[1] + start  # Must always start at "start"

  return(data.frame("t" = t, "xt" = path))
}
```
```{r Gaus}

head(df)
for(x in 1:5){
  df <- as.data.frame(gaussprocess())
  plot(df)
}

#ggplot(df)

```

```{r something}
# TK
```


## References
https://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html
https://www.r-bloggers.com/2012/04/gaussian-process-regression-with-r/
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4827276/